scripts/scratch_ingest/expression_writer.py

```diff
@@ -0,0 +1,416 @@
+        :param cluster_name: (String) name of cluster object
+        :param data_dir: (String) name out output dir
+        :param indexes: (List) start/stop index points to read from/to
+        """
+        start_pos, end_pos = indexes
+        print(f"reading {self.matrix_file_path} at index {start_pos}:{end_pos}")
+        with ExpressionWriter.open_file(self.matrix_file_path) as matrix_file:
+            current_pos = start_pos
+            matrix_file.seek(current_pos)
+            while current_pos < end_pos:
+                line = matrix_file.readline()
+                self.process_dense_line(line, matrix_cells, cluster_cells, cluster_name, data_dir)
+                current_pos += len(line)
+
+    def process_dense_line(self, line, matrix_cells, cluster_cells, cluster_name, data_dir):
+        """
+        Process a single line from a dense matrix and write gene-level data
+
+        :param matrix_cells: (List) cells from header line of dense matrix
+        :param cluster_cells: (List) cell names from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir (String) name out output dir
+        :param line: (String) single line from dense matrix
+        """
+        clean_line = line.rstrip().replace('"','')
+        raw_vals = re.split(COMMA_OR_TAB, clean_line)
+        gene_name = raw_vals.pop(0)
+        exp_vals = [round(float(val), self.precision) if float(val) != 0.0 else 0 for val in raw_vals]
+        filtered_expression = ExpressionWriter.filter_expression_for_cluster(
+            cluster_cells, matrix_cells, exp_vals
+        )
+        ExpressionWriter.write_gene_scores(cluster_name, gene_name, filtered_expression, data_dir)
+
+    @staticmethod
+    def filter_expression_for_cluster(cluster_cells, exp_cells, exp_scores):
+        """
+        Assemble a List of expression scores, filtered & ordered from a List of cluster cells
+        Will substitute 0 as a value for any cell not seen in the expression file
+
+        :param cluster_cells: (List) cluster cell names
+        :param exp_cells: (List) expression cell names
+        :param exp_scores: (List) expression values, in the same order as exp_cells
+        :return: (List) Expression values, ordered by cluster_cells
+        """
+        observed_exp = dict(zip(exp_cells, exp_scores))
+        return (observed_exp.get(cell, 0) for cell in cluster_cells)
+
+    @staticmethod
+    def write_gene_scores(cluster_name, gene_name, exp_values, data_dir):
+        """
+        Write a JSON array of expression values
+        Filename uses {cluster_name}--{gene_name}.json format
+
+        :param cluster_name: (String) Name of cluster
+        :param gene_name: (String) Name of gene
+        :param exp_values: (List) expression values
+        :param data_dir: (String) name out output dir
+        """
+        with gzip.open(f"{data_dir}/{cluster_name}--{gene_name}.json.gz", "wt") as file:
```

Comment:
I'm not entirely sure that it matters - I don't see anything about compression levels in their docs on [transcoding](https://cloud.google.com/storage/docs/transcoding) - so it could be that the answer is "yes".  We can always try and see what happens.

I did find that `gsutil cp` has a `-Z` flag which will _automatically gzip all files on upload_, so we may in fact want to write non-compressed and let `gsutil` handle it for us, as this will also set the appropriate `Content-Encoding` and `Cache-Control` headers.